src/serialiser/include/IoSerialiserYaS.hpp

```
@@ -148,11 +147,110 @@ struct IoSerialiser<YaS, T> {
     }
 };
 
+template<MapLike T>
+struct IoSerialiser<YaS, T> {
+    inline static constexpr uint8_t getDataTypeId() { return 203; }
+    constexpr static bool           serialise(IoBuffer &buffer, const ClassField &field, const T &value) noexcept {
+        using K                 = typename T::key_type;
+        using V                 = typename T::mapped_type;
+        const int32_t nElements = static_cast<int32_t>(value.size());
+        buffer.put(std::array<int32_t, 1>{ nElements }); // [ndims]{size}
+        buffer.put(nElements);                           // nElements
+
+        if constexpr (is_supported_number<K> || is_stringlike<K>) {
+            constexpr int entrySize = 17; // as an initial estimate
+            buffer.reserve_spare(static_cast<size_t>(nElements * entrySize + 9));
+            buffer.put(static_cast<uint8_t>(yas::getDataTypeId<K>())); // type-id
+            buffer.put(nElements);                                     // nElements
```

I would opt for not storing the number of elements redundantly, just takes up space and needs code to validate it again that should never be triggered when serialising correctly. Maybe also move the Types K and V before the data, that way, we can fail early in the deserialiser and do not have to deserialise all the keys just to flunk out on a value type mismatch.

---

I agree, but I copy-pasted this to be compatible with the existing opencmw-java implentation.